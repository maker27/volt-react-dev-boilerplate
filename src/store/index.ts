import { configureStore } from '@reduxjs/toolkit';
import invoicesReducer from './invoiceSlice';
import productReducer from './productSlice';
import customerReducer from './customerSlice';
import statusReducer from './statusSlice';

export const store = configureStore({
    reducer: {
        invoices: invoicesReducer,
        products: productReducer,
        customers: customerReducer,
        status: statusReducer
    }
});

export type RootState = ReturnType<typeof store.getState>;
export type AppDispatch = typeof store.dispatch;
