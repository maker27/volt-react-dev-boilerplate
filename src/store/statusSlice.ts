import { createSlice, PayloadAction } from '@reduxjs/toolkit';

export enum StatusType {
    warning,
    message,
    error
}

export type TStatus = { id: number; type: StatusType; value: string };

interface IStatusState {
    list: Array<TStatus>;
    loading: boolean;
}

export const initialState: IStatusState = {
    list: [],
    loading: false
};

export const statusSlice = createSlice({
    name: 'status',
    initialState,
    reducers: {
        addStatus: (
            state,
            action: PayloadAction<{ id: number; type: StatusType | undefined; value: string }>
        ) => {
            state.list.push({
                id: action.payload.id,
                type: action.payload.type || StatusType.message,
                value: action.payload.value
            });
        },
        removeStatus: (state, action: PayloadAction<number>) => {
            const removableStatusIndex = state.list.findIndex(({ id }) => id === action.payload);
            if (removableStatusIndex !== -1) state.list.splice(removableStatusIndex, 1);
        },
        setLoadAction: (state, action: PayloadAction<boolean>) => {
            state.loading = action.payload;
        }
    }
});

export const { addStatus, removeStatus, setLoadAction } = statusSlice.actions;

export default statusSlice.reducer;
