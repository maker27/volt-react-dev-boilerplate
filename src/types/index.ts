import React from 'react';
import { IInvoice, TInvoice } from './invoice';
import { IProduct, TProduct } from './product';
import { ICustomer, TCustomer } from './customer';
import { IInvoiceItem, TInvoiceItem } from './invoice-item';

export type { IInvoice, IProduct, ICustomer, IInvoiceItem, TCustomer, TProduct, TInvoice, TInvoiceItem };

export type TOnChangeEvent = React.ChangeEvent<HTMLInputElement>;

export type TObjectOf<T> = { [key: string]: T };

export type TFormDataSetter<T> = (value: T) => void;
