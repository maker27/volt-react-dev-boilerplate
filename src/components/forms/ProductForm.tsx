import React from 'react';
import { Form } from 'react-bootstrap';
import { TOnChangeEvent, TProduct, IProduct, TFormDataSetter } from '../../types';
import LabeledInput from './LabeledInput';

export default function ProductForm({
    product,
    setFormData
}: {
    product: IProduct;
    setFormData: TFormDataSetter<IProduct>;
}): JSX.Element {
    const setProductProp = (prop: keyof TProduct) => ({ target }: TOnChangeEvent) => {
        if (prop === 'price' && !/^(\d)*(\.)?([0-9]{0,2})?$/.test(target.value)) return;
        setFormData({ ...product, [prop]: target.value });
    };
    return (
        <Form>
            <LabeledInput
                id="form-product-name"
                label="Name"
                value={product.name}
                placeholder="Product Name"
                onChange={setProductProp('name')}
            />
            <LabeledInput
                id="form-product-price"
                label="Price (0.00)"
                value={String(product.price)}
                placeholder="00.00"
                onChange={setProductProp('price')}
            />
        </Form>
    );
}
