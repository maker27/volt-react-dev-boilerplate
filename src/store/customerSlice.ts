import { createSlice, PayloadAction } from '@reduxjs/toolkit';
import { ICustomer, TObjectOf } from '../types';
import { arrayToObject } from '../utils';

interface customerState {
    all: TObjectOf<ICustomer>;
    current: ICustomer;
}

const emptyCustomer = { id: 0, name: '', address: '', phone: '' };

const initialState: customerState = {
    all: {},
    current: emptyCustomer
};

export const customerSlice = createSlice({
    name: 'customer',
    initialState,
    reducers: {
        setCustomers: (state, action: PayloadAction<Array<ICustomer>>) => {
            state.all = arrayToObject<ICustomer>(action.payload);
        },
        addCustomer: (state, action: PayloadAction<ICustomer>) => {
            const customer = action.payload;
            state.all[customer.id] = customer;
        },
        editCustomer: (state, action: PayloadAction<ICustomer>) => {
            const customer = action.payload;
            state.all[customer.id] = customer;
        },
        deleteCustomer: (state, action: PayloadAction<ICustomer>) => {
            const customer = action.payload;
            delete state.all[customer.id];
        },
        setCurrentCustomer: (state, action: PayloadAction<ICustomer | undefined>) => {
            state.current = action.payload || emptyCustomer;
        }
    }
});

export const {
    setCustomers,
    addCustomer,
    editCustomer,
    deleteCustomer,
    setCurrentCustomer
} = customerSlice.actions;

export default customerSlice.reducer;
