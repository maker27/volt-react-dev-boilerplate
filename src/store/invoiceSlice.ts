import { createSlice, PayloadAction } from '@reduxjs/toolkit';
import { IInvoice, IInvoiceItem, TObjectOf } from '../types';
import { arrayToObject } from '../utils';

interface IInvoiceState {
    all: TObjectOf<IInvoice>;
    current: IInvoice;
    items: TObjectOf<IInvoiceItem>;
}

const emptyInvoice = {
    id: 0,
    customer_id: 0,
    discount: 0,
    total: 0
};

const initialState: IInvoiceState = {
    all: {},
    current: emptyInvoice,
    items: {}
};

export const invoiceSlice = createSlice({
    name: 'invoice',
    initialState,
    reducers: {
        setInvoices: (state, action: PayloadAction<Array<IInvoice>>) => {
            state.all = arrayToObject<IInvoice>(action.payload);
        },
        addInvoice: (state, action: PayloadAction<IInvoice>) => {
            const invoice = action.payload;
            state.all[invoice.id] = invoice;
            state.current = invoice;
        },
        editInvoice: (state, action: PayloadAction<IInvoice>) => {
            const invoice = action.payload;
            state.all[invoice.id] = invoice;
            state.current = invoice;
        },
        deleteInvoice: (state, action: PayloadAction<IInvoice>) => {
            const invoice = action.payload;
            delete state.all[invoice.id];
        },
        setCurrentInvoice: (state, action: PayloadAction<IInvoice | undefined>) => {
            state.current = action.payload || emptyInvoice;
        },
        setInvoiceItems: (state, action: PayloadAction<IInvoiceItem[]>) => {
            state.items = arrayToObject<IInvoiceItem>(action.payload, 'product_id');
        },
        addInvoiceItem: (state, action: PayloadAction<IInvoiceItem>) => {
            const invoiceItem = action.payload;
            state.items[invoiceItem.product_id] = invoiceItem;
        },
        deleteInvoiceItem: (state, action: PayloadAction<IInvoiceItem>) => {
            const invoiceItem = action.payload;
            delete state.items[invoiceItem.product_id];
        }
    }
});

export const {
    setInvoices,
    addInvoice,
    editInvoice,
    deleteInvoice,
    setCurrentInvoice,
    setInvoiceItems,
    addInvoiceItem,
    deleteInvoiceItem
} = invoiceSlice.actions;

export default invoiceSlice.reducer;
