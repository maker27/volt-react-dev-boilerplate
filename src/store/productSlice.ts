import { createSlice, PayloadAction } from '@reduxjs/toolkit';
import { IProduct, TObjectOf } from '../types';
import { arrayToObject } from '../utils';

interface productState {
    all: TObjectOf<IProduct>;
    current: IProduct;
}

const emptyProduct = { id: 0, name: '', price: 0 };

const initialState: productState = {
    all: {},
    current: emptyProduct
};

export const productSlice = createSlice({
    name: 'product',
    initialState,
    reducers: {
        setProducts: (state, action: PayloadAction<Array<IProduct>>) => {
            state.all = arrayToObject<IProduct>(action.payload);
        },
        addProduct: (state, action: PayloadAction<IProduct>) => {
            const product = action.payload;
            state.all[product.id] = product;
        },
        editProduct: (state, action: PayloadAction<IProduct>) => {
            const product = action.payload;
            state.all[product.id] = product;
        },
        deleteProduct: (state, action: PayloadAction<IProduct>) => {
            const product = action.payload;
            delete state.all[product.id];
        },
        setCurrentProduct: (state, action: PayloadAction<IProduct | undefined>) => {
            state.current = action.payload || emptyProduct;
        }
    }
});

export const {
    setProducts,
    addProduct,
    editProduct,
    deleteProduct,
    setCurrentProduct
} = productSlice.actions;

export default productSlice.reducer;
