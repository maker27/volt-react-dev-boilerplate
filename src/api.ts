import axios, { AxiosRequestConfig, AxiosResponse } from 'axios';
import { apiBaseUrl } from './assets/constants';
import { store } from './store';
import { ActionCreatorWithPayload } from '@reduxjs/toolkit';
import {
    IInvoice,
    IProduct,
    ICustomer,
    TCustomer,
    TProduct,
    TInvoice,
    IInvoiceItem,
    TInvoiceItem
} from './types';
import {
    addInvoice,
    addInvoiceItem,
    deleteInvoice,
    deleteInvoiceItem,
    editInvoice,
    editInvoiceItem,
    setInvoiceItems,
    setInvoices
} from './store/invoiceSlice';
import { addProduct, deleteProduct, editProduct, setProducts } from './store/productSlice';
import { addCustomer, deleteCustomer, editCustomer, setCustomers } from './store/customerSlice';
import { StatusType, addStatus, setLoadAction } from './store/statusSlice';

const server = axios.create({
    baseURL: apiBaseUrl
});

type TRequestData = { [key: string]: string | number };

const isInitialRequest = ({ method, url = '' }: AxiosRequestConfig) =>
    method === 'get' && ['customers', 'products', 'invoices'].includes(url);

class API {
    constructor() {
        server.interceptors.request.use((req: AxiosRequestConfig) => {
            if (!isInitialRequest(req)) API.setLoader(true);
            return req;
        });
        server.interceptors.response.use((res: AxiosResponse) => {
            if (!isInitialRequest(res.config)) API.setLoader(false);
            return res;
        });
    }

    private static setLoader(load: boolean) {
        store.dispatch(setLoadAction(load));
        return load;
    }

    private static setError(e: Error) {
        API.setLoader(false);
        store.dispatch(addStatus({ id: Date.now(), type: StatusType.error, value: e.message }));
        console.error(e);
    }

    private request<T>(
        /* eslint-disable-next-line @typescript-eslint/no-explicit-any */
        serverRequest: <T = any, R = AxiosResponse<T>>(
            url: string,
            config?: AxiosRequestConfig
        ) => Promise<R>,
        params: [url: string, data?: TRequestData],
        action: ActionCreatorWithPayload<T>
    ) {
        return serverRequest(...params)
            .then(({ data }) => {
                store.dispatch(action(data));
                return data;
            })
            .catch(e => API.setError(e));
    }

    private getRequest<T>(url: string, action: ActionCreatorWithPayload<T>) {
        return this.request(server.get, [url], action);
    }

    private postRequest<T>(url: string, data: TRequestData, action: ActionCreatorWithPayload<T>) {
        return this.request(server.post, [url, data], action);
    }

    private putRequest<T>(url: string, data: TRequestData, action: ActionCreatorWithPayload<T>) {
        return this.request(server.put, [url, data], action);
    }

    private deleteRequest<T>(url: string, action: ActionCreatorWithPayload<T>) {
        return this.request(server.delete, [url], action);
    }

    getInvoices() {
        return this.getRequest<IInvoice[]>('invoices', setInvoices);
    }

    getProducts() {
        return this.getRequest<IProduct[]>('products', setProducts);
    }

    getCustomers() {
        return this.getRequest<ICustomer[]>('customers', setCustomers);
    }

    getAllData() {
        return Promise.all([this.getInvoices(), this.getProducts(), this.getCustomers()]);
    }

    createCustomer(customerData: TCustomer) {
        this.postRequest<ICustomer>('customers', customerData, addCustomer);
    }

    editCustomer(customerId: number, customerData: TCustomer) {
        this.putRequest<ICustomer>(
            'customers/' + customerId,
            { ...customerData, customer_id: customerId },
            editCustomer
        );
    }

    deleteCustomer(customerId: number) {
        this.deleteRequest<ICustomer>('customers/' + customerId, deleteCustomer);
    }

    createProduct(productData: TProduct) {
        this.postRequest<IProduct>('products', productData, addProduct);
    }

    editProduct(productId: number, productData: TProduct) {
        this.putRequest<IProduct>(
            'products/' + productId,
            { ...productData, product_id: productId },
            editProduct
        );
    }

    deleteProduct(productId: number) {
        this.deleteRequest<IProduct>('products/' + productId, deleteProduct);
    }

    createInvoice(invoiceData: TInvoice) {
        return this.postRequest<IInvoice>('invoices', invoiceData, addInvoice);
    }

    editInvoice(invoiceId: number, invoiceData: Partial<TInvoice>) {
        this.putRequest<IInvoice>('invoices/' + invoiceId, invoiceData, editInvoice);
    }

    deleteInvoice(invoiceId: number) {
        this.deleteRequest<IInvoice>('invoices/' + invoiceId, deleteInvoice);
    }

    getInvoiceItems(invoiceId: number) {
        return this.getRequest<IInvoiceItem[]>('invoices/' + invoiceId + '/items', setInvoiceItems);
    }

    addInvoiceItem(invoiceItemData: TInvoiceItem) {
        return this.postRequest<IInvoiceItem>(
            'invoices/' + invoiceItemData.invoice_id + '/items',
            invoiceItemData,
            addInvoiceItem
        );
    }

    editInvoiceItem({ id, ...invoiceItemData }: IInvoiceItem) {
        this.putRequest<IInvoiceItem>(
            'invoices/' + invoiceItemData.invoice_id + '/items/' + id,
            { ...invoiceItemData },
            editInvoiceItem
        );
    }

    deleteInvoiceItem({ id, invoice_id }: IInvoiceItem) {
        this.deleteRequest<IInvoiceItem>('invoices/' + invoice_id + '/items/' + id, deleteInvoiceItem);
    }

    addNewInvoice(invoice: IInvoice, invoiceItems: TInvoiceItem[]) {
        API.setLoader(true);
        this.createInvoice(invoice).then(({ id }) => {
            Promise.all(
                Object.values(invoiceItems).map(invoiceItem =>
                    this.addInvoiceItem({ ...invoiceItem, invoice_id: id })
                )
            ).then(() => {
                API.setLoader(false);
            });
        });
    }
}

export default new API();
