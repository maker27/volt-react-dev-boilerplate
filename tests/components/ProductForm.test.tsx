import React from 'react';
import { render, screen, fireEvent, RenderResult } from '@testing-library/react';
import { Provider } from 'react-redux';
import { store } from '../../src/store';
import ProductForm from '../../src/components/forms/ProductForm';
import { IProduct } from '../../src/types';
import { addProduct, setCurrentProduct } from '../../src/store/productSlice';
import { testProductsList } from '../data';

const setFormData = jest.fn();

const renderForm = (product: IProduct): RenderResult => {
    const setProduct = (newProduct?: IProduct) => store.dispatch(setCurrentProduct(newProduct));
    return render(
        <Provider store={store}>
            <ProductForm product={product} setFormData={setProduct} />
        </Provider>
    );
};

const getProduct = (productId: number): IProduct => {
    const product = store.getState().products.all[productId];
    expect(product).not.toBeUndefined();
    return product as IProduct;
};

const getCurrentProduct = (): IProduct => {
    const product = store.getState().products.current;
    expect(product).not.toBeUndefined();
    return product as IProduct;
};

describe('ProductForm', () => {
    const addedProduct = testProductsList[0];
    beforeAll(() => {
        store.dispatch(addProduct(addedProduct));
    });

    test('renders form', () => {
        const product = getProduct(addedProduct.id);
        const { asFragment } = renderForm(product);
        expect(asFragment()).toMatchSnapshot();
    });

    test('changes name field', () => {
        let product = getProduct(addedProduct.id);
        const newValue = 'Test Name';

        const { asFragment, rerender } = renderForm(product);
        let textInput = screen.getByLabelText('Name') as HTMLInputElement;
        expect(textInput.value).toBe(addedProduct.name);
        fireEvent.change(textInput, { target: { value: newValue } });

        textInput = screen.getByLabelText('Name') as HTMLInputElement;
        product = getCurrentProduct();
        expect(product).toMatchInlineSnapshot(`
            Object {
              "id": ${addedProduct.id},
              "name": "${newValue}",
              "price": ${addedProduct.price},
            }
        `);

        rerender(
            <Provider store={store}>
                <ProductForm product={product} setFormData={setFormData} />
            </Provider>
        );
        expect(textInput.value).toBe(newValue);
        expect(asFragment()).toMatchSnapshot();
    });
});
