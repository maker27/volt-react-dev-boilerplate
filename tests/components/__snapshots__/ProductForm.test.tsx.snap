// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ProductForm changes name field 2`] = `
<DocumentFragment>
  <form
    class=""
  >
    <div
      class="mb-3 "
    >
      <label
        class="fw-bold form-label"
        for="form-product-name"
      >
        Name
      </label>
      <input
        class="form-control"
        id="form-product-name"
        maxlength="255"
        placeholder="Product Name"
        type="text"
        value="Test Name"
      />
    </div>
    <div
      class="mb-3 "
    >
      <label
        class="fw-bold form-label"
        for="form-product-price"
      >
        Price (0.00)
      </label>
      <input
        class="form-control"
        id="form-product-price"
        maxlength="255"
        placeholder="00.00"
        type="text"
        value="23.39"
      />
    </div>
  </form>
</DocumentFragment>
`;

exports[`ProductForm renders form 1`] = `
<DocumentFragment>
  <form
    class=""
  >
    <div
      class="mb-3 "
    >
      <label
        class="fw-bold form-label"
        for="form-product-name"
      >
        Name
      </label>
      <input
        class="form-control"
        id="form-product-name"
        maxlength="255"
        placeholder="Product Name"
        type="text"
        value="Product 101"
      />
    </div>
    <div
      class="mb-3 "
    >
      <label
        class="fw-bold form-label"
        for="form-product-price"
      >
        Price (0.00)
      </label>
      <input
        class="form-control"
        id="form-product-price"
        maxlength="255"
        placeholder="00.00"
        type="text"
        value="23.39"
      />
    </div>
  </form>
</DocumentFragment>
`;
